packages/app/src/hooks/useServiceMap.tsx

```diff
@@ -0,0 +1,295 @@
+  metadata,
+  samplingFactor,
+}: {
+  source: TSource;
+  dateRange: [Date, Date];
+  traceId?: string;
+  metadata: Metadata;
+  samplingFactor: number;
+}) {
+  // Don't sample if we're looking for a specific trace
+  const effectiveSamplingLevel = traceId ? 1 : samplingFactor;
+
+  const baseCTEConfig = {
+    from: source.from,
+    connection: source.connection,
+    dateRange,
+    timestampValueExpression: source.timestampValueExpression,
+    filters: [
+      // Sample a subset of traces, for performance in the following join
+      {
+        type: 'sql' as const,
+        condition: `cityHash64(${source.traceIdExpression}) % ${effectiveSamplingLevel} = 0`,
+      },
+      // Optionally filter for a specific trace ID
+      ...(traceId
+        ? [
+            {
+              type: 'sql' as const,
+              condition: SqlString.format('?? = ?', [
+                source.traceIdExpression,
+                traceId,
+              ]),
+            },
+          ]
+        : []),
+    ],
+    select: [
+      {
+        valueExpression: source.traceIdExpression ?? 'TraceId',
+        alias: 'traceId',
+      },
+      {
+        valueExpression: source.spanIdExpression ?? 'SpanId',
+        alias: 'spanId',
+      },
+      {
+        valueExpression: source.serviceNameExpression ?? 'ServiceName',
+        alias: 'serviceName',
+      },
+      {
+        valueExpression: source.parentSpanIdExpression ?? 'ParentSpanId',
+        alias: 'parentSpanId',
+      },
+      {
+        valueExpression: source.statusCodeExpression ?? 'StatusCode',
+        alias: 'statusCode',
+      },
+    ],
+  };
+
+  const [serverCTE, clientCTE] = await Promise.all([
+    renderChartConfig(
+      {
+        ...baseCTEConfig,
+        filters: [
+          ...baseCTEConfig.filters,
+          {
+            type: 'sql',
+            condition: `${source.spanKindExpression} IN ('Server', 'Consumer')`,
+          },
+        ],
+        where: '',
+      },
+      metadata,
+    ),
+    renderChartConfig(
+      {
+        ...baseCTEConfig,
+        filters: [
+          ...baseCTEConfig.filters,
+          {
+            type: 'sql',
+            condition: `${source.spanKindExpression} IN ('Client', 'Producer')`,
+          },
+        ],
+        where: '',
+      },
+      metadata,
+    ),
+  ]);
+
+  // Left join to support services which receive requests from clients that are not instrumented.
+  // Ordering helps ensure stable graph layout.
+  return chSql`
+    WITH 
+      ServerSpans AS (${serverCTE}),
+      ClientSpans AS (${clientCTE})
+    SELECT
+      ServerSpans.serviceName AS serverServiceName,
+      ServerSpans.statusCode AS serverStatusCode,
+      ClientSpans.serviceName AS clientServiceName,
+      count(*) * ${{ Int64: effectiveSamplingLevel }} as requestCount
+    FROM ServerSpans
+      LEFT JOIN ClientSpans
+        ON ServerSpans.traceId = ClientSpans.traceId
+        AND ServerSpans.parentSpanId = ClientSpans.spanId
```

Comment:
Perf: not within the scope of this ticket. I'm concerned about the performance implications here, since the default schema doesn't have indexes on traceId or spanId.